This is an exploration of a dataset of 50,000 used car listings scraped from 'eBay Kleinanzeign', a section of the German eBay website.  We will clean the data and provide high level analysis.  

In [1]:
import pandas as pd
import numpy as np
autos = pd.read_csv("autos.csv", encoding="Latin-1")

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5      2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   
6      2016-03-20 17:55:21  VW_Golf_III_GT_Special_Electronic_Green_Metall...   
7      2016-03-16 18:55:19                               Golf_IV_1.9_TDI_90PS   
8      2016-03-22 16:51:34                                         Seat_Arosa   
9      2016-03-16 13:47:02          Renault_Megane_Scenic_1.6e_RT_Klimaanlage   
10     2016-03-15 01:41:36                       VW_Golf_Tuning_in_siber/grau   
11     2016-03-16 18:45:34                         Mercedes_A140_Motorschaden   
12     2016-03-31 19:48:22  Smart_smart_fortwo_coupe_softouch_pure_MHD_Pan...   
13     2016-03-23 10:48:32                                 Audi_A3_1.6_tuning   
14     2016-03-23 11:50:46  Renault_Clio_3__Dynamique_1.2__16_V;_viele_Ver...   
15     2016-04-01 12:06:20        Corvette_C3_Coupe_T_Top_Crossfire_Injection   
16     2016-03-16 14:59:02                                Opel_Vectra_B_Kombi   
17     2016-03-29 11:46:22                          Volkswagen_Scirocco_2_G60   
18     2016-03-26 19:57:44               Verkaufen_mein_bmw_e36_320_i_touring   
19     2016-03-17 13:36:21        mazda_tribute_2.0_mit_gas_und_tuev_neu_2018   
20     2016-03-05 19:57:31  Audi_A4_Avant_1.9_TDI_*6_Gang*AHK*Klimatronik*...   
21     2016-03-06 19:07:10                      Porsche_911_Carrera_4S_Cabrio   
22     2016-03-28 20:50:54                               MINI_Cooper_S_Cabrio   
23     2016-03-10 19:55:34   Peugeot_Boxer_2_2_HDi_120_Ps_9_Sitzer_inkl_Klima   
24     2016-04-03 11:57:02                         BMW_535i_xDrive_Sport_Aut.   
25     2016-03-21 21:56:18  Ford_escort_kombi_an_bastler_mit_ghia_ausstattung   
26     2016-04-03 22:46:28                                Volkswagen_Polo_Fox   
27     2016-03-27 18:45:01             Hat_einer_Ahnung_mit_Ford_Galaxy_HILFE   
28     2016-03-19 21:56:19                                      MINI_Cooper_D   
29     2016-04-02 12:45:44  Mercedes_Benz_E_320_T_CDI_Avantgarde_DPF7_Sitz...   
...                    ...                                                ...   
49970  2016-03-21 22:47:37  c4_Grand_Picasso_mit_Automatik_Leder_Navi_Temp...   
49971  2016-03-29 14:54:12                                         W.Lupo_1.0   
49972  2016-03-26 22:25:23          Mercedes_Benz_Vito_115_CDI_Extralang_Aut.   
49973  2016-03-27 05:32:39                   Mercedes_Benz_SLK_200_Kompressor   
49974  2016-03-20 10:52:31  Golf_1_Cabrio_Tuev_Neu_viele_Extras_alles_eing...   
49975  2016-03-27 20:51:39          Honda_Jazz_1.3_DSi_i_VTEC_IMA_CVT_Comfort   
49976  2016-03-19 18:56:05  Audi_80_Avant_2.6_E__Vollausstattung!!_Einziga...   
49977  2016-03-31 18:37:18                        Mercedes_Benz_C200_Cdi_W203   
49978  2016-04-04 10:37:14                        Mercedes_Benz_E_200_Classic   
49979  2016-03-20 18:38:40                      Volkswagen_Polo_1.6_TDI_Style   
49980  2016-03-12 10:55:54                            Ford_Escort_Turnier_16V   
49981  2016-03-15 09:38:21             Opel_Astra_Kombi_mit_Anhaengerkupplung   
49982  2016-03-29 18:51:08            Skoda_Fabia_4_Tuerer_Bj:2004__85.000Tkm   
49983  2016-03-06 12:43:04                                      Ford_focus_99   
49984  2016-03-31 22:48:48  Student_sucht_ein__Anfaengerauto___ab_2000_BJ_...   
49985  2016-04-02 16:38:23                          Verkaufe_meinen_vw_vento!   
49986  2016-04-04 20:46:02  Chrysler_300C_3.0_CRD_DPF_Automatik_Voll_Ausst...   
49987  2016-03-22 20:47:27  Audi_A3_Limousine_

In [3]:
print(autos.info())
print(autos.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
dateCrawled            50000 non-null object
name                   50000 non-null object
seller                 50000 non-null object
offerType              50000 non-null object
price                  50000 non-null object
abtest                 50000 non-null object
vehicleType            44905 non-null object
yearOfRegistration     50000 non-null int64
gearbox                47320 non-null object
powerPS                50000 non-null int64
model                  47242 non-null object
odometer               50000 non-null object
monthOfRegistration    50000 non-null int64
fuelType               45518 non-null object
brand                  50000 non-null object
notRepairedDamage      40171 non-null object
dateCreated            50000 non-null object
nrOfPictures           50000 non-null int64
postalCode             50000 non-null int64
lastSeen               50000 non-null obj

Dataset will have to be cleaned particularly as it relates to date/time entries.  Additionally, a number of the columns stored as strings will have to be converted to integers or floats (eg. odometer).  Further, there are a number of null values which we will either replace or eliminate.  Finally, we'll clean up the column names.

In [4]:
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [5]:
new_column_names = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'odometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen']
autos.columns = new_column_names
autos.head(3)

date_crawled                                        name  seller  \
0  2016-03-26 17:47:46            Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1  2016-04-04 13:38:56  BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   
2  2016-03-26 18:57:24                  Volkswagen_Golf_1.6_United  privat   

  offer_type   price   abtest vehicle_type  registration_year    gearbox  \
0    Angebot  $5,000  control          bus               2004    manuell   
1    Angebot  $8,500  control    limousine               1997  automatik   
2    Angebot  $8,990     test    limousine               2009    manuell   

   power_ps   model   odometer  registration_month fuel_type       brand  \
0       158  andere  150,000km                   3       lpg     peugeot   
1       286     7er  150,000km                   6    benzin         bmw   
2       102    golf   70,000km                   7    benzin  volkswagen   

  unrepaired_damage           ad_created  nr_of_pictures  postal_code  \
0              nein  2016-03-26 00:00:00               0        79588   
1              nein  2016-04-04 00:00:00               0        71034   
2              nein  2016-03-26 00:00:00               0        35394   

             last_seen  
0  2016-04-06 06:45:54  
1  2016-04-06 14:45:08  
2  2016-04-06 20:15:37

Renamed a few columns so they are more descriptive (eg. notRepairedDamage changed to unrepaired_damage) and changed all column names to snakecase from camelcase.  

In [6]:
autos['nr_of_pictures'].value_counts()
autos['unrepaired_damage'].value_counts()
autos['vehicle_type'].value_counts()
autos['offer_type'].value_counts()
autos['seller'].value_counts()

privat        49999
gewerblich        1
Name: seller, dtype: int64

Need to clean:
    1) Dates
    2) price
    3) odometer
    
Can likely get rid of nr_of_pictures since they are all 0's.  Can also probably get rid of seller and offer_type since each column has the same values with just 1 exception.

In [7]:
# price:
autos["price"] = autos["price"].str.replace("$","")
autos["price"] = autos["price"].str.replace(",","").astype(int)

In [8]:
# odometer:
autos["odometer"] = autos["odometer"].str.replace("km","")
autos["odometer"] = autos["odometer"].str.replace(",","").astype(int)
autos.rename({"odometer" : "odometer_km"}, axis=1, inplace=True)

In [9]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price   abtest vehicle_type  registration_year  \
0  privat    Angebot   5000  control          bus               2004   
1  privat    Angebot   8500  control    limousine               1997   
2  privat    Angebot   8990     test    limousine               2009   
3  privat    Angebot   4350  control   kleinwagen               2007   
4  privat    Angebot   1350     test        kombi               2003   

     gearbox  power_ps   model  odometer_km  registration_month fuel_type  \
0    manuell       158  andere       150000                   3       lpg   
1  automatik       286     7er       150000                   6    benzin   
2    manuell       102    golf        70000                   7    benzin   
3  automatik        71  fortwo        70000                   6    benzin   
4    manuell         0   focus       150000                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [10]:
autos["odometer_km"].unique().shape
autos["odometer_km"].describe()
autos["odometer_km"].value_counts().sort_index(ascending=False).head()

150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
Name: odometer_km, dtype: int64

In [11]:
autos["price"].unique().shape
autos["price"].describe()
autos["price"].value_counts().sort_index(ascending=True).head(50)
autos = autos[autos["price"].between(25,250000)]

Explored the columns odometer_km and price.  odometer_km only has 13 unique values, each one a round number.  Additionally, there are more than 30,000 values equal to 150,000 which indicates that the value is actually a cut off.  So 150,000 actually means more than 150,000 km.  

On the other hand, price has a very large number of values.  The values range from 0 to 1 million.  There are more than 1,000 cars with a price of 0.  It is unlikely that any legitimate sale would take place for less than 25 Euro so we will exclude these listings.  Similarly, it is unlikely that anyone would buy a car on eBay for more than 250,000 Euro.  So we have removed these rows as well.  We are left with roughly 48,000 of the original 50,000 rows.

In [12]:
autos["date_crawled"] = autos["date_crawled"].str[:10]
autos["date_crawled"].sort_index().value_counts(normalize=True, dropna=False)

/dataquest/system/env/python3/lib/python3.4/site-packages/ipykernel/__main__.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if __name__ == '__main__':


2016-04-03    0.038578
2016-03-20    0.037792
2016-03-21    0.037296
2016-03-12    0.036965
2016-03-14    0.036655
2016-04-04    0.036510
2016-03-07    0.036055
2016-04-02    0.035518
2016-03-28    0.034856
2016-03-19    0.034753
2016-03-15    0.034277
2016-03-29    0.034133
2016-04-01    0.033761
2016-03-30    0.033740
2016-03-08    0.033264
2016-03-09    0.033058
2016-03-22    0.032934
2016-03-11    0.032562
2016-03-26    0.032272
2016-03-23    0.032272
2016-03-10    0.032231
2016-03-31    0.031838
2016-03-17    0.031590
2016-03-25    0.031528
2016-03-27    0.031094
2016-03-16    0.029502
2016-03-24    0.029419
2016-03-05    0.025346
2016-03-13    0.015650
2016-03-06    0.014058
2016-04-05    0.013045
2016-03-18    0.012901
2016-04-06    0.003163
2016-04-07    0.001385
Name: date_crawled, dtype: float64

In [13]:
autos["ad_created"] = autos["ad_created"].str[:10]
autos["ad_created"].value_counts(normalize=True, dropna=False).sort_index()

/dataquest/system/env/python3/lib/python3.4/site-packages/ipykernel/__main__.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if __name__ == '__main__':


2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
2015-12-30    0.000021
2016-01-03    0.000021
2016-01-07    0.000021
2016-01-10    0.000041
2016-01-13    0.000021
2016-01-14    0.000021
2016-01-16    0.000021
2016-01-22    0.000021
2016-01-27    0.000062
2016-01-29    0.000021
2016-02-01    0.000021
2016-02-02    0.000041
2016-02-05    0.000041
2016-02-07    0.000021
2016-02-08    0.000021
2016-02-09    0.000021
2016-02-11    0.000021
2016-02-12    0.000041
2016-02-14    0.000041
2016-02-16    0.000021
2016-02-17    0.000021
2016-02-18    0.000041
2016-02-19    0.000062
2016-02-20    0.000041
2016-02-21    0.000062
                ...   
2016-03-09    0.033140
2016-03-10    0.031941
2016-03-11    0.032892
2016-03-12    0.036779
2016-03-13    0.017015
2016-03-14    0.035290
2016-03-15    0.034009
2016-03-16    0.029998
2016-03-17    0.031259
2016-03-18    0.013583
2016-03-19    0.033637
2016-03-20    0.037854
2016-03-21 

In [14]:
autos["last_seen"] = autos["last_seen"].str[:10]
autos["last_seen"].value_counts(normalize=True, dropna=False).sort_index()

/dataquest/system/env/python3/lib/python3.4/site-packages/ipykernel/__main__.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if __name__ == '__main__':


2016-03-05    0.001075
2016-03-06    0.004321
2016-03-07    0.005417
2016-03-08    0.007360
2016-03-09    0.009634
2016-03-10    0.010626
2016-03-11    0.012384
2016-03-12    0.023775
2016-03-13    0.008869
2016-03-14    0.012632
2016-03-15    0.015878
2016-03-16    0.016456
2016-03-17    0.028075
2016-03-18    0.007319
2016-03-19    0.015816
2016-03-20    0.020612
2016-03-21    0.020591
2016-03-22    0.021398
2016-03-23    0.018607
2016-03-24    0.019764
2016-03-25    0.019165
2016-03-26    0.016808
2016-03-27    0.015568
2016-03-28    0.020860
2016-03-29    0.022307
2016-03-30    0.024767
2016-03-31    0.023858
2016-04-01    0.022886
2016-04-02    0.024871
2016-04-03    0.025181
2016-04-04    0.024519
2016-04-05    0.124953
2016-04-06    0.221584
2016-04-07    0.132065
Name: last_seen, dtype: float64

Sorting the date_crawled, the ad_created and last_seen columns by date and calculating the percentage of rows with each date revealed some interesing information.  With regards to the date_crawled, it appears as if all of the instances were crawled between 3/6/16 and 4/7/16.  There is a fairly even distribution with roughly 3% per day although the most recent days (after 4/4/16) seem to have fewer cars (less than 1.2% each).  

The ad_created column has a much earlier starting point.  The earliest ad was created in June of 2015.  There are a small number of cars that apparently have not sold in a number of months.  However, the bulk of the ads were created in the same timeframe as the date_crawled range.  

Finally, the last_seen column has the same range as the date_crawled column.  However, there are a much higher percentage of last_seen over the last few days of the range.  Each of 4/5/16, 4/6/16 and 4/7/16 represent more than 12% of the distribution.  Presumably, a large percentage of sellers are checking the listing fairly regularly.  

In [15]:
autos["registration_year"].describe()
autos["registration_year"].value_counts().sort_index(ascending=True).head(50)

1000     1
1001     1
1111     1
1800     2
1910     3
1927     1
1929     1
1931     1
1934     2
1937     4
1938     1
1939     1
1941     2
1943     1
1948     1
1950     1
1951     2
1952     1
1953     1
1954     2
1955     2
1956     4
1957     2
1958     4
1959     6
1960    22
1961     6
1962     4
1963     8
1964    12
1965    17
1966    22
1967    26
1968    26
1969    19
1970    37
1971    26
1972    33
1973    23
1974    24
1975    18
1976    21
1977    21
1978    42
1979    34
1980    82
1981    28
1982    41
1983    51
1984    51
Name: registration_year, dtype: int64

Once again, there appears to be some errors in the registration_year column.  The values range from 1000 to 9999.  Of course, it is unlikely that any cars were actually registered before 1930 and it is impossible that cars were registered in the future (ie after 2017 assuming the auto manufacturers sell 2017 cars in 2016).  We'll remove rows before 1930 and after 2017.

In [16]:
autos = autos[autos["registration_year"].between(1930, 2017)]

In [17]:
autos.describe()

price  registration_year      power_ps    odometer_km  \
count   47874.000000       47874.000000  47874.000000   47874.000000   
mean     5893.453399        2003.344195    117.367068  125864.456699   
std      8459.151950           7.369081    187.987953   39611.319840   
min        25.000000        1931.000000      0.000000    5000.000000   
25%      1250.000000        1999.000000     73.000000  125000.000000   
50%      3000.000000        2003.000000    107.500000  150000.000000   
75%      7499.000000        2008.000000    150.000000  150000.000000   
max    250000.000000        2017.000000  17700.000000  150000.000000   

       registration_month  nr_of_pictures   postal_code  
count        47874.000000         47874.0  47874.000000  
mean             5.804675             0.0  51044.467268  
std              3.676607             0.0  25750.882483  
min              0.000000             0.0   1067.000000  
25%              3.000000             0.0  30826.000000  
50%              6.000000             0.0  49762.000000  
75%              9.000000             0.0  71706.000000  
max             12.000000             0.0  99998.000000

In [18]:
autos["brand"].value_counts(normalize=True, dropna=False)

volkswagen        0.212809
bmw               0.108744
opel              0.108117
mercedes_benz     0.095939
audi              0.086143
ford              0.069913
renault           0.047688
peugeot           0.029640
fiat              0.025901
seat              0.018611
skoda             0.016167
nissan            0.015353
mazda             0.015311
smart             0.014329
citroen           0.014120
toyota            0.012721
hyundai           0.010005
sonstige_autos    0.009421
volvo             0.009045
mini              0.008710
mitsubishi        0.008105
honda             0.007917
kia               0.007144
alfa_romeo        0.006642
suzuki            0.005870
porsche           0.005807
chevrolet         0.005640
chrysler          0.003530
dacia             0.002674
daihatsu          0.002507
jeep              0.002235
subaru            0.002068
land_rover        0.002068
saab              0.001629
daewoo            0.001546
jaguar            0.001525
trabant           0.001379
r

In [19]:
brands = ['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault', 'peugeot', 'fiat']

In [20]:
brand_price_dictionary = {}
brand_year_dictionary = {}
brand_mileage_dictionary = {}
for b in brands:
    temp = autos[autos['brand'] == b]
    mean_price = temp['price'].mean()
    mean_registration = temp['registration_year'].mean()
    mean_odometer = temp['odometer_km'].mean()
    brand_price_dictionary[b] = mean_price
    brand_year_dictionary[b] = mean_registration
    brand_mileage_dictionary[b] = mean_odometer
bp_series = pd.Series(brand_price_dictionary)
by_series = pd.Series(brand_year_dictionary)
bm_series = pd.Series(brand_mileage_dictionary)
df = pd.DataFrame(bp_series, columns=['mean_price'])
df['mean_year'] = by_series
df['mean_mileage'] = bm_series

In [21]:
print(df)

                mean_price    mean_year   mean_mileage
audi           9272.919981  2004.470660  129507.759457
bmw            8271.865924  2003.307914  132743.949289
fiat           2795.586290  2003.725806  117314.516129
ford           3738.486705  2003.193606  124291.903197
mercedes_benz  8559.826475  2002.396473  131040.714130
opel           2971.688563  2002.804096  129461.939722
peugeot        3078.063425  2004.020437  127269.203665
renault        2452.131406  2003.488830  128219.448095
volkswagen     5368.206125  2002.985179  128947.290931


Grouping the top brands (any brand with more than 2% of the rows), we calculated the average price, registration year and odometer.  Interestingly, there was very little difference among the average registration year and average km.  The oldest car on average was Mercedes Benz followed closely by BMW, both averaging early 2002 while the newest was Audi at mid 2004.  Fiats on average had the fewest kms at 117,000 while BMWs had the most average kms at just under 133,000.

Price, on the other hand, was quite variable ranging from a low of just 2,452 Euros for Renault to a high of 9,273 Euros for Audis.  Despite being among the oldest cars listed, both BMW and Mercedes also averaged very high prices at more than 8,000 Euros each.  Similarly, despite having the lowest average kms, Fiats had the second lowest average price at 2,795 Euros.  Apparently, brand matters much more than age of the car or mileage.    

In [22]:
autos.head()

date_crawled                                               name  seller  \
0   2016-03-26                   Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1   2016-04-04         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   
2   2016-03-26                         Volkswagen_Golf_1.6_United  privat   
3   2016-03-12  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...  privat   
4   2016-04-01  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...  privat   

  offer_type  price   abtest vehicle_type  registration_year    gearbox  \
0    Angebot   5000  control          bus               2004    manuell   
1    Angebot   8500  control    limousine               1997  automatik   
2    Angebot   8990     test    limousine               2009    manuell   
3    Angebot   4350  control   kleinwagen               2007  automatik   
4    Angebot   1350     test        kombi               2003    manuell   

   power_ps   model  odometer_km  registration_month fuel_type       brand  \
0       158  andere       150000                   3       lpg     peugeot   
1       286     7er       150000                   6    benzin         bmw   
2       102    golf        70000                   7    benzin  volkswagen   
3        71  fortwo        70000                   6    benzin       smart   
4         0   focus       150000                   7    benzin        ford   

  unrepaired_damage  ad_created  nr_of_pictures  postal_code   last_seen  
0              nein  2016-03-26               0        79588  2016-04-06  
1              nein  2016-04-04               0        71034  2016-04-06  
2              nein  2016-03-26               0        35394  2016-04-06  
3              nein  2016-03-12               0        33729  2016-03-15  
4              nein  2016-04-01               0        39218  2016-04-01

In [23]:
# Translating German to English
mapping_dict = {
    'privat': 'private',
    'gewerblich' : 'commercial',
    'Angebot' : 'offer',
    'limousine' : 'sedan',
    'kleinwagen': 'compact',
    'kombi' : 'station wagon',
    'bus' : 'bus',
    'cabrio' : 'cabrio',
    'coupe' : 'coupe',
    'suv' : 'suv',
    'andere' : 'other',
    'automatik': 'automatic',
    'manuell': 'manual',
    'lpg' : 'lpg',
    'benzin' : 'gas',
    'diesel' : 'diesel',
    'lpg' : 'lpg',
    'cng' : 'cng',
    'hybrid' : 'hybrid',
    'elektro' : 'electric',
    'nein' : 'no',
    'ja' : 'yes'
}
translate = ['seller', 'offer_type', 'vehicle_type', 'gearbox', 
            'fuel_type', 'unrepaired_damage']
for t in translate:
    autos[t] = autos[t].map(mapping_dict)

In [24]:
autos['fuel_type'].value_counts()

gas         29069
diesel      14265
lpg           662
cng            71
hybrid         37
electric       18
other          15
Name: fuel_type, dtype: int64

In [25]:
autos.head()

date_crawled                                               name   seller  \
0   2016-03-26                   Peugeot_807_160_NAVTECH_ON_BOARD  private   
1   2016-04-04         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  private   
2   2016-03-26                         Volkswagen_Golf_1.6_United  private   
3   2016-03-12  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...  private   
4   2016-04-01  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...  private   

  offer_type  price   abtest   vehicle_type  registration_year    gearbox  \
0      offer   5000  control            bus               2004     manual   
1      offer   8500  control          sedan               1997  automatic   
2      offer   8990     test          sedan               2009     manual   
3      offer   4350  control        compact               2007  automatic   
4      offer   1350     test  station wagon               2003     manual   

   power_ps   model  odometer_km  registration_month fuel_type       brand  \
0       158  andere       150000                   3       lpg     peugeot   
1       286     7er       150000                   6       gas         bmw   
2       102    golf        70000                   7       gas  volkswagen   
3        71  fortwo        70000                   6       gas       smart   
4         0   focus       150000                   7       gas        ford   

  unrepaired_damage  ad_created  nr_of_pictures  postal_code   last_seen  
0                no  2016-03-26               0        79588  2016-04-06  
1                no  2016-04-04               0        71034  2016-04-06  
2                no  2016-03-26               0        35394  2016-04-06  
3                no  2016-03-12               0        33729  2016-03-15  
4                no  2016-04-01               0        39218  2016-04-01

In [35]:
# What are the most common brand/model combinations?
model_counts = []
brands = ['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault', 'peugeot', 'fiat']
for b in brands:
    temp = autos[autos['brand']== b]
    models = temp['model'].unique().tolist()
    for m in models:
        temp2 = temp[temp['model'] == m]
        count = len(temp2)
        model_counts.append([b, m, count])
model_counts.sort(key=lambda x: x[2], reverse=True)

In [37]:
model_counts[0:20]

[['volkswagen', 'golf', 3848],
 ['bmw', '3er', 2653],
 ['volkswagen', 'polo', 1657],
 ['opel', 'corsa', 1644],
 ['opel', 'astra', 1384],
 ['volkswagen', 'passat', 1371],
 ['audi', 'a4', 1247],
 ['mercedes_benz', 'c_klasse', 1155],
 ['bmw', '5er', 1138],
 ['mercedes_benz', 'e_klasse', 966],
 ['audi', 'a3', 858],
 ['audi', 'a6', 807],
 ['ford', 'focus', 780],
 ['ford', 'fiesta', 745],
 ['volkswagen', 'transporter', 684],
 ['renault', 'twingo', 637],
 ['peugeot', '2_reihe', 608],
 ['mercedes_benz', 'a_klasse', 574],
 ['opel', 'vectra', 550],
 ['bmw', '1er', 530]]

In [42]:
# How does mileage impact price?
autos['odometer_km'].value_counts()
low_mileage_price = autos.loc[autos['odometer_km'] < 50001, 'price'].mean()
medium_mileage_price = autos.loc[autos['odometer_km'] < 100001, 'price'].mean()
high_mileage_price = autos.loc[autos['odometer_km'] > 100001, 'price'].mean()
print(low_mileage_price, medium_mileage_price, high_mileage_price)

14573.561552595314 11353.198353356296 4086.794028522977


In [45]:
# How does damage impact price?
damage_price = autos.loc[autos['unrepaired_damage'] == 'yes', 'price'].mean()
undamage_price = autos.loc[autos['unrepaired_damage'] == 'no', 'price'].mean()
print(damage_price, undamage_price)

2231.141842445114 7064.271336951159
